diff --git a/dist/client/app/composables/copyCode.js b/dist/client/app/composables/copyCode.js
index 43354d8fa4d6892c84bd937571d3af351d701705..69782357fe012ef659b67a52b627d712d0a9f42f 100644
--- a/dist/client/app/composables/copyCode.js
+++ b/dist/client/app/composables/copyCode.js
@@ -11,11 +11,13 @@ export function useCopyCode() {
                     return;
                 }
                 const isShell = /language-(shellscript|shell|bash|sh|zsh)/.test(parent.className);
-                let text = '';
-                sibling
-                    .querySelectorAll('span.line:not(.diff.remove)')
-                    .forEach((node) => (text += (node.textContent || '') + '\n'));
-                text = text.slice(0, -1);
+                const ignoredNodes = ['.vp-copy-ignore', '.diff.remove'];
+                // Clone the node and remove the ignored nodes
+                const clone = sibling.cloneNode(true);
+                clone
+                    .querySelectorAll(ignoredNodes.join(','))
+                    .forEach((node) => node.remove());
+                let text = clone.textContent || '';
                 if (isShell) {
                     text = text.replace(/^ *(\$|>) /gm, '').trim();
                 }
